iceoryx2-bb/container/src/slotmap.rs

```diff
@@ -0,0 +1,639 @@
+        ///
+        ///  * [`RelocatableSlotMap::init()`] must be called once before
+        ///
+        pub unsafe fn get(&self, key: SlotMapKey) -> Option<&T> {
+            self.get_impl(key)
+        }
+
+        /// Returns a mutable reference to the value stored under the given key. If there is no
+        /// such key, [`None`] is returned.
+        ///
+        /// # Safety
+        ///
+        ///  * [`RelocatableSlotMap::init()`] must be called once before
+        ///
+        pub unsafe fn get_mut(&mut self, key: SlotMapKey) -> Option<&mut T> {
+            self.get_mut_impl(key)
+        }
+
+        /// Insert a value and returns the corresponding [`SlotMapKey`]. If the container is full
+        /// [`None`] is returned.
+        ///
+        /// # Safety
+        ///
+        ///  * [`RelocatableSlotMap::init()`] must be called once before
+        ///
+        pub unsafe fn insert(&mut self, value: T) -> Option<SlotMapKey> {
+            self.insert_impl(value)
+        }
+
+        /// Insert a value at the specified [`SlotMapKey`] and returns true.  If the provided key
+        /// is out-of-bounds it returns `false` and adds nothing. If there is already a value
+        /// stored at the `key`s index, the value is overridden with the provided value.
+        ///
+        /// # Safety
+        ///
+        ///  * [`RelocatableSlotMap::init()`] must be called once before
+        ///
+        pub unsafe fn insert_at(&mut self, key: SlotMapKey, value: T) -> bool {
+            self.insert_at_impl(key, value)
+        }
+
+        /// Removes a value at the specified [`SlotMapKey`]. If there was no value corresponding
+        /// to the [`SlotMapKey`] it returns false, otherwise true.
+        ///
+        /// # Safety
+        ///
+        ///  * [`RelocatableSlotMap::init()`] must be called once before
+        ///
+        pub unsafe fn remove(&mut self, key: SlotMapKey) -> bool {
+            self.remove_impl(key)
+        }
+
+        /// Returns the number of stored values.
+        pub fn len(&self) -> usize {
+            self.len_impl()
+        }
+
+        /// Returns the capacity.
+        pub fn capacity(&self) -> usize {
+            self.capacity_impl()
+        }
+
+        /// Returns true if the container is empty, otherwise false.
+        pub fn is_empty(&self) -> bool {
+            self.is_empty_impl()
+        }
+
+        /// Returns true if the container is full, otherwise false.
+        pub fn is_full(&self) -> bool {
+            self.is_full_impl()
+        }
+    }
+}
+
+/// A compile-time fixed-size, shared memory compatible [`FixedSizeSlotMap`].
+#[repr(C)]
+#[derive(Debug)]
+pub struct FixedSizeSlotMap<T, const CAPACITY: usize> {
+    state: RelocatableSlotMap<T>,
+    _idx_to_data: MaybeUninit<[usize; CAPACITY]>,
+    _idx_to_data_free_list: MaybeUninit<[FreeListEntry; CAPACITY]>,
+    _data: MaybeUninit<[Option<T>; CAPACITY]>,
+    _data_next_free_index: MaybeUninit<[usize; CAPACITY]>,
+}
```

Comment:
I think this is something that could easily be created with a proc macro. No so sure about implementing `default` via the proc macro. That might be more difficult but at least the struct members would automatically be kept in sync.

Reply:
You are maybe right. Sooner or later we have to revisit the relocatable container concept and then we introduce something like this here. Currently, I would stick to the current approach since it is used also in the other containers - but I agree, I also do not like it at all!

Reply:
I'm not sure if there is currently another way to do it, except having a proc macro that calculates the memory size and adds a `buffer: [u8; N]` member to the struct. Anyway, without a proc macro the current way is less cumbersome and with the proc macro it doesn't really matter since it's hidden from the developer.